# Clean Code Study - Week3
#### 2023-02-03. Fri. written by HT.LEE
- - -
## Ch7. Error Handling
+ 소개
    - Clean Code와 오류 처리는 불가분의 관계
    - 무언가가 잘못될 가능성은 늘 존재한다.
    - 무언가가 잘못되면 그것을 바로 잡을 책임은 프로그래머에게 있다.
    - 오류처리 코드로 인해 프로그램의 논리를 이해하기 어려워진다면?  
        -> not clean.
    - 따라서 <b>논리적 흐름과 오류처리 흐름을 분리</b>해야한다.

+ 오류 코드보단 예외를 사용하자
    - C와 같은, 과거의 프로그래밍 언어에서는 예외를 제공하지 않아서 에러코드를 반환하고 if-else로 예외를 처리했다.
    - <a href="https://android.googlesource.com/kernel/lk/+/dima/for-travis/include/errno.h"> errno.h에서 정의된 예외코드 </a>
    
    - List 7-1. DeviceController.java
    ```Java
    public class DeviceController{
        // ...
        DeviceHandle handle = getHandle(DEV1);
        // Check the state of the Device
        if (handle != DeviceHandle.INVALID){
            // Save the device status to the record field
            retrieveDeviceRecord(handle);
            if (record.getStatus() != DEVICE_SUSPENDED){
                pauseDevice(handle);
                clearDeviceWorkQueue(handle);
                closeDevice(handle);
            }else{
                logger.log("Device suspended. Unable to shut down.");
            }
        }else{
            logger.log("Invalid handle for: " + DEV1.toString());
        }
        //...
    }
    ```

    - 이 방법은 중첩된 if-else 목록들을 포함할 가능성이 커져서 코드가 복잡해진다.
    - 오류가 발생하면 오류코드로 확인하여 처리하는 것보다는, 예외를 던지는 편이 낫다.

+ try-catch-finally 구문부터 쓰자
    - C++, Java, JS 등 대부분의 경우 try-catch-finally
    - Python의 경우, try-except-else-finally
    - 참고 : Stack unwinding
        * <a href="https://learn.microsoft.com/en-us/cpp/cpp/exceptions-and-stack-unwinding-in-cpp?view=msvc-170">
            Exceptions and Stack Unwinding in C++
        </a>

    - 예외가 발생할 가능성이 있는 로직의 실행 <b>범위를 지정</b>할 수 있다.
    - 트랜잭션과 유사하다.
    - 프로그램 상태를 일관적으로 유지하는 것이 목표이다.
    - 따라서, 예외가 발생할 코드에는 <b>try-catch-finally 구문부터 작성</b>하는 것이 좋다.
    - try-catch 구조로 범위를 정의하면, TDD를 사용해 필요한 나머지 논리를 추가할 수 있다.

+ Use Unchecked Exceptions
    - 결론적으로, 확인된 예외를 사용하지 말자고 합의함.
    - 옛날 Java에서는 메서드를 선언할 때, 메서드가 반환할 예외를 모두 열거했다.(확인된 예외)
        * 메서드가 반환하는 예외는 메서드 유형의 일부였음.
        * 코드가 메서드를 사용하는 방식이 메서드 선언과 불일치하면 아예 컴파일이 안 됨.

    - 확인된 예외를 사용하는 것은 OCP 원칙을 위반한다.
        * Top()
        * Middle1() throws IOException
        * Middle2() throws IOException
        * Bottom() throws IOException
        * 위와 같이 4개의 함수가 있다. Top()->Middle1()->Middle2()->Bottom()호출, Top에서 최종적으로 try-catch 처리
        * 만약, Bottom()에서 IOException이 아닌 다른 타입의 Exception을 호출하는 것으로 메서드 유형을 바꾼다면? Top과 Bottom 사이의 모든 메서드들을 수정해야함.
        
+ Provide Context with Exceptions
    - 예외를 던질 때, 오류 메시지에 정보를 담아 함께 던져라!
        * 실패한 연산 이름, 실패한 유형도 언급해라.
        * 가능하다면 발생한 오류 메시지에 대해 로그를 찍도록 해라.
    - 그러면 오류가 발생한 원인과 위치를 찾기가 쉬워진다.

+ Define Exception Classes in Terms of a Caller's Needs
    - 애플리케이션에서 오류를 정의할 때 프로그래머는 <b>오류를 잡아내는 방법</b>에 대해 가장 고민을 많이 해야한다!
    <i>After</i>
    ```Java
    ACMEPort port = new ACMEPort(12);

    try {
        port.open();
    }catch (DeviceResponseException e){
        reportPortError(e);
        logger.log("Device response exception", e);
    }catch (ATM1212UnlockedException e){
        reportPortError(e);
        logger.log("Unlock exception", e);
    }catch (GMXError e){
        reportPortError(e);
        logger.log("Device reponse exception", e);
    }finally{
        //...
    }
    ```
    - 위 코드의 대다수 상황에서 예외에 대응하는 방식이 예외 유형과 무관하게 거의 동일하다는 점을 주목해라.
        1. 오류를 기록한다.
        2. 프로그램을 계속 수행해도 좋을지 확인한다.

    <i>After</i>
    ```Java
    LocalPort port = new LocalPort(12);

    try {
        port.open();
    }catch (PortDeviceFailure e){
        reportPortError(e);
        logger.log(e.getMessage(), e);
    }finally{
        //...
    }
    
    public class LocalPort{
        private ACMEPort innerPort;

        public LocalPort(int portNumber){
            innerPort = new ACMEPort(portNumber);
        }

        public void open(){
            try{
                innerPort.open();
            }catch (DeviceResponseException e){
                throw new PortDeviceFailure(e);
            }catch (ATM1212UnlockedException e){
                throw new PortDeviceFailure(e);
            }catch (GMXError e){
                throw new PortDeviceFailure(e);
            }
        }
        //...
    }
    ```

    외부 라이브러리의 ACMEPort에서 정의된 여러 exception들을 단일 규격의 exception으로 처리하기 위해 ACMEPort를 LocalPort라는 wrapping 클래스로 감쌌다. 이러한 감싸기 기법은 매우 유용하다.
    '
    <외부 API를 사용할 때 Wrapping하는 기법의 장점>
    1. 외부 라이브러리와 프로그램 사이에서의 의존성이 크게 줄어든다.
    2. Wrapper 클래스에서 외부 API를 호출하는 대신, 테스트 코드를 넣어주는 방법으로 프로그램을 테스트하기도 쉬워진다.
    3. 특정 업체가 만든 API 설계 방식에 얽매이지 않는다.

+ Define the Normal Flow
    - 예외를 던지고, 코드 위에 핸들러를 정의해 중단된 계산을 처리하는 방식이 만능은 아니다. (중단이 적합하지 않을 수가 있다)
    
    - 아래 코드는, 비용 청구 애플리케이션에서 총계를 계산하는 코드이다.
    <i>Before</i>
    ```Java
    try {
        MealExpenses expenses = expenseReportDAO.getMeals(employee.getID());
        m_total += expenses.getTotal();
    }catch (MeaelExpensesNotFound e){
        m_total += getMealPerDiem();
    }
    ```
    - " m_total += (...) " 이 try와 catch 두 곳에서 반복됨. 특수 케이스에 대해 m_total을 따로 더하는 것이 논리적 흐름을 방해한다.
    
    <i>After</i>
    ```Java
    MealExpenses expenses = expenseReportDAO.getMeals
    m_total += expenses.getTotal();
    // ...
    public class PerDiemMealExpenses implements MealExpenses{
        public int getTotal(){
            // 기본값으로 일일 기본 식비 (meal per diem)을 반환함.
        }
    }
    ```
    - 청구한 식비가 없다면 MealPerDiem을 반환하는 "특수 케이스" 객체를 생성한다.
    - 위와 같은 방식을 SPEICAL CASE PATTERN이라고 부른다.
    - 이렇게 정상적인 작업흐름을 규정하면, 클래스나 객체가 예외적인 상황을 캡슐화해서 처리할 수 있다.

+ Don't Return Null
    ```Java
    public void registerItem(Item item){
        if (item != null){
            ItemRegistry registry = persistentStore.getItemRegistry();
            if (registry != null){
                Item existing = registry.getItem(itemGetID());
                if (existing.getBillingPeriod().hasRetailOwner()){
                    existing.register(item);
                }
            }
        }
    }
    ```
    - 만약 persistentStore가 null이라면? NullPointerException 발생!
    이 코드는 null 확인이 누락되서 문제가 되는 것이 아니라, 너무 많아서 문제가 되는 코드이다.

    - 가급적 메서드에서 null을 반환하지말고, 예외를 던지거나 special-case object를 만들어서 반환해라.

    <i>Before</i>
    ```Java
    List<Employee> employees = getEmployees();
    if (employees != null){
        for (Employee e : employees){
            totalPay += e.getPay();
        }
    }
    ```
    <i>After</i>
    ```Java
    List<Employee> employees = getEmployees();
    for (Employee e : employees){
            totalPay += e.getPay();
    }
    public List<Employee> getEmployees(){
        if (... there are no employees ...)
            return Collections.emptyList();
    }
    ```    

+ Don't Pass Null
    - null을 메서드의 인자로 넣는 것은 더 심각한 문제이다.
    - 메서드에 null을 전달하는 것은 문제 발생 위치를 찾기 힘들게 한다.
    
    ```Java
    public class MetricsCalculator
    {
        public double xProjection(Point p1, Point p2){
            return (p2.x - p1.x) * 1.5;
        }
    }
    ```
    - 위의 메서드에 누군가가 null을 인자로 넣으면 어떻게 될까?    
    - 해결방안
        1. Invalid arguments에 대한 exception을 만들어 던져라.
        2. 아예 null 자체를 못 받도록 만들어라. (Kotlin에는 언어적 지원이 있음)
        3. assert 구문을 사용해라.

+ 결론
    - 클린 코드는 높은 가독성과 안정성을 요구한다.
    - 가독성과 안정성은 상충되는 목표가 아님.
    - 프로그램 논리와 오류 처리를 확실히 분리하여, 한 번에 한 작업에만 (논리부분이든, 얘외처리 부분이든) 집중할 수 있게 해라.

- - -
+ 심화 학습
    - Special Case Pattern 다른 사례 살펴볼 것!